ledger/src/blockstore_metrics.rs

```diff
@@ -516,7 +512,43 @@ pub(crate) fn report_rocksdb_write_perf(metric_header: &'static str) {
 /// A struct that holds the current status of RocksDB perf sampling.
 pub struct PerfSamplingStatus {
     // The number of RocksDB operations since the last perf sample.
-    pub(crate) op_count: AtomicUsize,
+    op_count: AtomicUsize,
+    // The timestamp of the latest operation with perf stats collection.
+    last_sample_time_ms: AtomicU64,
+}
+
+impl PerfSamplingStatus {
+    fn should_sample(&self, sample_count_interval: usize) -> bool {
+        if sample_count_interval == 0 {
+            return false;
+        }
+
+        // Rate-limiting based on the number of samples.
+        if self.op_count.fetch_add(1, Ordering::Relaxed) < sample_count_interval {
+            return false;
+        }
+        self.op_count.store(0, Ordering::Relaxed);
+
+        // Rate-limiting based on the time duration.
+        let current_time_ms = timestamp();
+        let old_time_ms = self.last_sample_time_ms.load(Ordering::Relaxed);
+        if old_time_ms + (PERF_SAMPLING_MIN_DURATION.as_millis() as u64) < current_time_ms {
```

Comment:
this line looks wrong

Reply:
If the last sampling time (`old_time`) is at least `PERF_SAMPLING_MIN_DURATION.as_millis()` older than the `current_time_ms`, then we should proceed instead of returning false.

`<` should be replaced by `>` instead.

Reply:
Fixed!